node/src/service/nonce_monitor.rs

```diff
@@ -0,0 +1,199 @@
+use std::time::Duration;
+
+use codec::Decode;
+use creditcoin_node_runtime::AccountId;
+use jsonrpc_core::{futures::channel::mpsc, Failure, Response, Success};
+use sc_client_api::Backend;
+use sc_service::{Arc, RpcHandlers, RpcSession};
+use sp_runtime::{app_crypto::Ss58Codec, offchain::OffchainStorage};
+use substrate_prometheus_endpoint::Registry;
+use thiserror::Error;
+
+use super::FullBackend;
+
+#[derive(Clone, Copy, PartialEq, Debug, Error)]
+enum ErrorKind {
+	#[error("Serde error")]
+	Serde,
+	#[error("RPC error")]
+	Rpc,
+	#[error("Codec error")]
+	Codec,
+}
+
+#[derive(Debug, Error)]
+#[error("{kind}: {value}")]
+struct Error {
+	kind: ErrorKind,
+	#[source]
+	value: Box<dyn std::error::Error + Send>,
+}
+
+macro_rules! error_fn {
+	($($fn_id: ident : $kind: ident),+ $(,)?) => {
+		impl Error {
+			$(
+				fn $fn_id(err: impl std::error::Error + Send + 'static) -> Self {
+					Self::new(ErrorKind::$kind, err)
+				}
+			)+
+		}
+	};
+}
+
+error_fn!(serde: Serde, rpc: Rpc, codec: Codec);
+
+impl Error {
+	fn new(kind: ErrorKind, err: impl std::error::Error + Send + 'static) -> Self {
+		Self { kind, value: Box::new(err) }
+	}
+}
+
+#[derive(Clone, PartialEq, Debug, Error)]
+#[error("{message}")]
+struct AdhocError {
+	message: String,
+}
+
+fn adhoc(message: impl Into<String>) -> AdhocError {
+	AdhocError { message: message.into() }
+}
+
+async fn rpc_request(
+	handlers: &RpcHandlers,
+	request: &str,
+) -> Result<jsonrpc_core::serde_json::Value, Error> {
+	let (tx, _rx) = mpsc::unbounded();
+	let session = RpcSession::new(tx);
+
+	let response = handlers
+		.rpc_query(&session, request)
+		.await
+		.ok_or_else(|| Error::rpc(adhoc("empty response")))?;
+
+	let response: Response = jsonrpc_core::serde_json::from_str(&response).map_err(Error::serde)?;
+
+	let result = match response {
+		Response::Single(out) => match out {
+			jsonrpc_core::Output::Success(Success { result, .. }) => result,
+			jsonrpc_core::Output::Failure(Failure { error, .. }) => return Err(Error::rpc(error)),
+		},
+		Response::Batch(_) => {
+			unreachable!("we don't send any batch requests, so we cannot receive batch responses")
+		},
+	};
+
+	Ok(result)
```

Comment:
```suggestion
match response {
		Response::Single(out) => match out {
			jsonrpc_core::Output::Success(Success { result, .. }) => Ok(result),
			jsonrpc_core::Output::Failure(Failure { error, .. }) => Err(Error::rpc(error)),
		},
		Response::Batch(_) => {
			unreachable!("we don't send any batch requests, so we cannot receive batch responses")
		},
	}
```

Reply:
I'm realizing I missed some of the suggested changes here, in the interest of getting the testnet release out (and to avoid the CI time + reapprovals) I'll merge this now and follow up with a PR to clean this up